dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java

```diff
@@ -57,12 +61,27 @@ public List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWork
           "Not enough workers in the cluster %d workers in the cluster but %d required",
           blockWorkerInfos.size(), count));
     }
-    HASH_PROVIDER.refresh(blockWorkerInfos, mNumVirtualNodes);
-    List<BlockWorkerInfo> workers = HASH_PROVIDER.getMultiple(fileId, count);
+    List<WorkerIdentity> workerIdentities = blockWorkerInfos.stream()
+        .map(BlockWorkerInfo::getIdentity)
+        .collect(Collectors.toList());
+    mHashProvider.refresh(workerIdentities, mNumVirtualNodes);
+    List<WorkerIdentity> workers = mHashProvider.getMultiple(fileId, count);
     if (workers.size() != count) {
       throw new ResourceExhaustedException(String.format(
           "Found %d workers from the hash ring but %d required", workers.size(), count));
     }
-    return workers;
+    ImmutableList.Builder<BlockWorkerInfo> builder = ImmutableList.builder();
+    // todo(bowen): this is quadratic complexity. examine if it's worthwhile to replace
+    //  with an indexed map if #workers is huge
+    for (WorkerIdentity worker : workers) {
```

Comment:
will there be a chance where the worker identites are not consistent with workers in blockWorkerInfo? In such case what will happen?

Reply:
`mHashProvider.refresh` should update the provider's internal state to be the same as the workers in blockWorkerInfo. In cases where the provider's internal cache of the workers has not expired (TTL is 1s), it could be true that the provider is using an stale view of the cluster. In this case, this one particular call will fail. After the TTL is up, the provider should be updated to use the latest worker list.